File: app/assets/javascripts/models/chart_filter.js
Cibi.ChartFilter = DS.Model.extend({
	fieldName: DS.attr('string'),
	formatAs: DS.attr('string'),
	comparisonOperator: DS.attr('number'),
	fieldValues: DS.attr('string'),
	exclude: DS.attr('boolean'),
	disabled: DS.attr('boolean'),

	fieldDataType: DS.attr('string'),
	dateRange: DS.attr('boolean'),
	upperRange: DS.attr('date'),
	lowerRange: DS.attr('date'),
	referenceDirection: DS.attr('string'),
  	referenceCount: DS.attr('number'),
  	referenceUnit: DS.attr('string'),
  	referenceDateToday: DS.attr('string'),
  	referenceDate: DS.attr('date'),
  	displayName: DS.attr('string'),
  	isGlobal: DS.attr('boolean'),

	chart: DS.belongsTo('Cibi.Chart'),
	user: DS.belongsTo('Cibi.User'),
	
	operatorsMap:  [
		{display: "", value: null},
		{display: "in", value: 0},
		{display: "less than", value: 1},
		{display: "greater than", value: 2},
		{display: "not in", value: 3},
		{display: "not null", value: 4},
		{display: "max", value: 6},
		{display: "min", value: 7},
	],

	values: function() {
		var field_values = this.get('fieldValues');
		if(!field_values) {
			return [];
		}
		return JSON.parse(field_values);
	}.property('fieldValues'),

	values_str: function() {
		var vals = this.get('values')
		if(this.get('comparisonOperator') == 4){
			return 'NOT NULL'
		}
		vals = _.map(vals, function(v) {
			return displayTick(v);
		})
		return vals.join(',');
	}.property('values'),

	operator: function() {
		var co = this.get('comparisonOperator');
		if(co == 4 || co == 6 || co == 7){
			return 'is';
		}else if(co == 5){
			return 'range';
		}
		if(co === undefined || co === null ) {
			return;
		}
		var operator = this.get('operatorsMap').filter(function(d) {
			return d.value == co;
		});
		return operator[0].display;
	}.property('comparisonOperator', 'isLoaded'),

	comparison: function(){
		var co = this.get('comparisonOperator');
		switch(co) {
			case 4:
				return "NOT NULL";
				break;
			case 6:
				return "MAX";
				break;
			case 7:
				return "MIN";
				break;
		}
	}.property('comparisonOperator', 'isLoaded'),

	apply: function(data) {
		var obj = this;
		if(obj.get('isLoaded')) {
			var field_name = obj.get('fieldName');
			var values = obj.get('values');
			var comparisonOperator = obj.get('comparisonOperator');
			switch(comparisonOperator) {
				case 0:
					data = _.filter(data, function(d) {
						if(!values) {
							return true;
						}
						return values.indexOf(d[field_name]) != -1;
					});
					break;
				case 1:
					data = _.filter(data, function(d) {
						var val = toFloat(d[field_name]);
						return val < toFloat(values);
					});
					break;	
				case 2:
					data = _.filter(data, function(d) {
						var val = toFloat(d[field_name]);
						return val > toFloat(values);
					});
					break;	
				case 3:
					data = _.filter(data, function(d) {
						if(!values) {
							return true;
						}
						return values.indexOf(d[field_name]) == -1;
					});
					break;	
				case 4:
					data = _.filter(data, function(d) {						
						return d[field_name]
					});
					break;	
				default:			
					alert('operator default');
			}			
		}
		return data;
	},

});

File: app/models/chart_filter.rb
class ChartFilter < AuditedModel
    attr_accessible :field_name, :format_as, :comparison_operator, :field_values, :exclude, :chart_id, :disabled, 
    				:field_data_type, :date_range, :upper_range, :lower_range, :reference_direction, 
    				:reference_count, :reference_unit, :reference_date_today, :reference_date, :display_name, 
    				:is_global, :user_id

    belongs_to :chart
    
    
	COMPARISON_OPERATOR = [
		"in",
		"lt",
		"gt",
		"not in",
		"not null"
	]
	
	def self.operator_type(operator)
		return self::COMPARISON_OPERATOR.index(operator)
	end

	def self.operator(operator_type)
		return self::COMPARISON_OPERATOR[operator_type]
	end

	def self.create_chart_filter(params)
		chart_filter = ChartFilter.create(params)
		chart_filter
	end

	def get_months_index(month_names)
		arr = []
		month_names.each do |name|
			arr << Date::MONTHNAMES.index(name)
		end
		arr
	end

	def get_condition
		operator = self.get_operator
		values = self.field_values ? self.get_filter_values : nil
		format_as = self.format_as
		arr = []

		# {"fieldName" => self.field_name, "formatAs" => self.format_as,"dataType" => "string",
		# "comparision" => operator, "value" => values}					
		if format_as == 'Month'
			values = get_months_index(values)
			arr << {"fieldName" => self.field_name, "formatAs" => format_as,"dataType" => "string",
		"comparision" => operator, "value" => values}
		elsif format_as == 'Month Year'
			vals = []
 			values.each do |val|
 				value = val.split(' ')
 				vals << "#{value[0]} "+("%02d" % Date::ABBR_MONTHNAMES.index(value[1]))				
 			end
 			values=vals
 			arr <<
 		         {"fieldName" => self.field_name, "formatAs" => format_as,"dataType" => "string",
			 "comparision" => self.get_operator,"value" => values}	
		else
			if(self.comparison_operator == 5 && self.date_range == true)
				get_range_condition(arr)
			elsif(self.comparison_operator == 5 && self.date_range == false)
				get_reference_condition(arr)
			else
				values = values
				arr << {"fieldName" => self.field_name, "formatAs" => format_as,"dataType" => "string",
				"comparision" => operator, "value" => values}	
			end

		end		
		arr
	end

	def get_range_condition(arr)
	        arr << {"fieldName" => self.field_name, "formatAs" => self.format_as,"dataType" => "string",
			 "comparision" => ">","value" => self.lower_range}
		 	arr << {"fieldName" => self.field_name, "formatAs" => self.format_as,"dataType" => "string",
			 "comparision" => "<","value" => self.upper_range}			 
	end

	def get_reference_condition(arr)
		reference_date = self.reference_date_today ? Date.today : self.reference_date
		if self.reference_direction == 'previous'		
			start_date = reference_date - self.reference_count.send(self.reference_unit.downcase)
			end_date = reference_date	
		else
			start_date = reference_date
			end_date = reference_date + self.reference_count.send(self.reference_unit.downcase)	
		end
    		arr << {"fieldName" => self.field_name, "formatAs" => self.format_as,"dataType" => "string",
			 "comparision" => ">","value" => start_date}
		 	arr << {"fieldName" => self.field_name, "formatAs" => self.format_as,"dataType" => "string",
			 "comparision" => "<","value" => end_date}
	end

	def get_operator
		case self.comparison_operator
		when 0
			return "IN"
		when 1
			return "<"
		when 2
			return ">"
		when 3
			return "NOT IN"
		when 4 
			return "IS NOT NULL"
		when 6
			return "MAX"
		when 7
			return "MIN"
		else
			return "IN" #default
		end
	end

	def get_filter_values
		unless self.field_values
			return nil			
		end
		val = JSON.parse(self.field_values, quirks_mode: true)
		operator = self.comparison_operator
		if operator == 1 || operator == 2
			val =val.to_i
		end
		val
	end
end
